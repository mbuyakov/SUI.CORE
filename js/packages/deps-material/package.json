{
  "name": "@sui/deps-material",
  "version": "0.0.1",
  "type": "commonjs",
  "dependencies": {
    "@emotion/react": "11.10.5",
    "@emotion/styled": "11.10.5",
    "@mui/icons-material": "5.14.3",
    "@mui/material": "5.14.4",
    "material-ui-popup-state": "5.0.9"
  }
}
