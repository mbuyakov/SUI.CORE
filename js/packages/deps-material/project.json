{
  "name": "deps-material",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/deps-material/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nx/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/deps-material",
        "main": "packages/deps-material/src/index.ts",
        "tsConfig": "packages/deps-material/tsconfig.lib.json",
        "assets": ["packages/deps-material/*.md"],
        "updateBuildableProjectDepsInPackageJson": true
      }
    },
    "publish": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs deps-material {args.ver} {args.tag}"
      },
      "dependsOn": ["build"]
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/deps-material/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "packages/deps-material/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": ["type:deps"]
}
