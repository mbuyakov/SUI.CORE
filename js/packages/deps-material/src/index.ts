// Need to correct indexed suggestion in IDEA
export * from '@mui/material';
export * as MuiIcons from '@mui/icons-material';
export * as MuiColors from "@mui/material/colors";
