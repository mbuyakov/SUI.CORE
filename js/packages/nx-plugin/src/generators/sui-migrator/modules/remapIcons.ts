import {ImportDeclaration, NamedImports, StringLiteral} from "typescript";
import {logWithPrefix} from "../../../utils/logger";
// eslint-disable-next-line @nx/enforce-module-boundaries
import {astReplace} from "@sui/lib-typescript-ast";

const SUFFIX = "MuiIcons.";

export function remapIcons(content: string): string {
  const iconsImports: ImportDeclaration[] = [];

  content = astReplace(content, "ImportDeclaration:has(StringLiteral[value=/@mui.icons-material.*/])", (node: ImportDeclaration) => {
    iconsImports.push(node);
    return "import {MuiIcons} from \"@sui/deps-material\";";
  });

  const nameMap = {};

  iconsImports.forEach(importDeclaration => {
    // Default import from subpath. Replace icon name to name from last part of path
    if (importDeclaration.importClause.name) {
      nameMap[importDeclaration.importClause.name.text] = SUFFIX + (importDeclaration.moduleSpecifier as StringLiteral).text.split("/").pop();
    } else {
      (importDeclaration.importClause.namedBindings as NamedImports).elements.forEach(importSpecifier => {
        nameMap[importSpecifier.name.text] = SUFFIX + importSpecifier.name.text;
      });
    }
  });

  Object.keys(nameMap).forEach(key => {
    logWithPrefix("remapIcons", `Replace ${key} to ${nameMap[key]}`);

    content = astReplace(content, `Identifier[name="${key}"]`, () => {
      return nameMap[key];
    });
  });

  return content;
}
