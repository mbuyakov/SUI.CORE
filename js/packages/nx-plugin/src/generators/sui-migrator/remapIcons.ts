import {factory, ImportDeclaration, NamedImports, StringLiteral} from "typescript";
import {logRemap, printNode} from "./util";
import {tsquery} from "./tsqeury";

const SUFFIX = "MuiIcons.";

export function remapIcons(content: string): string {
  const ast = tsquery.ast(content);

  const iconsImports: ImportDeclaration[] = [];
  iconsImports.push(...tsquery<ImportDeclaration>(ast, "ImportDeclaration:has(StringLiteral[value=/@mui.icons-material.*/])"));
  iconsImports.push(...tsquery<ImportDeclaration>(ast, "ImportDeclaration:has(StringLiteral[value=/@material-ui.icons.*/])"));

  if (!iconsImports.length) {
    return content;
  }

  content = tsquery.remove(content, "ImportDeclaration:has(StringLiteral[value=/@mui.icons-material.*/])", () => true, true);
  content = tsquery.remove(content, "ImportDeclaration:has(StringLiteral[value=/@material-ui.icons.*/])", () => true, true);

  const nameMap = {};

  iconsImports.forEach(importDeclaration => {
    // Default import from subpath. Replace icon name to name from last part of path
    if (importDeclaration.importClause.name) {
      nameMap[importDeclaration.importClause.name.text] = SUFFIX + (importDeclaration.moduleSpecifier as StringLiteral).text.split("/").pop();
    } else {
      (importDeclaration.importClause.namedBindings as NamedImports).elements.forEach(importSpecifier => {
        nameMap[importSpecifier.name.text] = SUFFIX + importSpecifier.name.text;
      });
    }
  });

  Object.keys(nameMap).forEach(key => {
    logRemap("remapIcons", `Replace ${key} to ${nameMap[key]}`);

    content = tsquery.jsxReplace(content, `Identifier[name="${key}"]`, () => {
      return nameMap[key];
    });
  });

  let importReplaced = false;
  content = tsquery.replace(content, "ImportDeclaration", (importDeclaration: ImportDeclaration) => {
    if (importReplaced) {
      return;
    }

    importReplaced = true;
    return printNode(
        factory.createImportDeclaration(
          undefined,
          factory.createImportClause(
            false,
            undefined,
            factory.createNamedImports([
              factory.createImportSpecifier(
                false,
                undefined,
                factory.createIdentifier("MuiIcons")
              )
            ])
          ),
          factory.createStringLiteral("@sui/deps-material")
        )
      )
        .replace(/{ /g, "{")
        .replace(/ }/g, "}")
      + "\n"
      + importDeclaration.getText();
  });

  return content;
}
