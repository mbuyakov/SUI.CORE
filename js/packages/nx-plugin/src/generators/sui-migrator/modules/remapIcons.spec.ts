import {remapIcons} from "./remapIcons";

describe("remap icons", () => {

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should leave as is", () => {
    const consoleLog = jest.spyOn(console, "log");
    const source = `
import {MuiIcons} from "@sui/deps-material";

const a = <MuiIcons.Menu attr="1">2</MuiIcons.Menu>;
const b = <MuiIcons.Menu attr="1"/>;

const c = <MuiIcons.DeleteForever attr="1">2</MuiIcons.DeleteForever>;
const d = <MuiIcons.DeleteForever attr="1"/>;

const f = <DrawerVisibleButton
            edge="start"
            disableRotate={isMobile}
            icon={isMobile && <MuiIcons.Menu/>}
          />;
const g = MuiIcons.Menu as any;
    `.trim();
    expect(remapIcons(source)).toBe(source);
    expect(consoleLog).not.toBeCalled();
  });

  it("should replace default import from subpath", () => {
    const consoleLog = jest.spyOn(console, "log");
    expect(remapIcons(`
import MenuIcon from "@mui/icons-material/Menu";

const a = <MenuIcon attr="1">2</MenuIcon>;
const b = <MenuIcon attr="1"/>;
const c = <DrawerVisibleButton
            edge="start"
            disableRotate={isMobile}
            icon={isMobile && <MenuIcon/>}
          />;
const d = MenuIcon as any;
    `.trim())).toMatchSnapshot();
    expect(consoleLog).toBeCalledTimes(1);
  });

  it("should replace named import", () => {
    const consoleLog = jest.spyOn(console, "log");
    expect(remapIcons(`
import {DeleteForever} from "@mui/icons-material";

const a = <DeleteForever attr="1">2</DeleteForever>;
const b = <DeleteForever attr="1"/>;
const c = <DrawerVisibleButton
            edge="start"
            disableRotate={isMobile}
            icon={isMobile && <DeleteForever/>}
          />;
const d = DeleteForever as any;
    `.trim())).toMatchSnapshot();
    expect(consoleLog).toBeCalledTimes(1);
  });
});
