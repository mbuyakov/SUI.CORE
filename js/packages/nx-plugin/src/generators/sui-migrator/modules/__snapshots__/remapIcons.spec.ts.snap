// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remap icons should replace default import from subpath 1`] = `
"import {MuiIcons} from \\"@sui/deps-material\\";

const a = <MuiIcons.Menu attr=\\"1\\">2</MuiIcons.Menu>;
const b = <MuiIcons.Menu attr=\\"1\\"/>;
const c = <DrawerVisibleButton
            edge=\\"start\\"
            disableRotate={isMobile}
            icon={isMobile && <MuiIcons.Menu/>}
          />;
const d = MuiIcons.Menu as any;"
`;

exports[`remap icons should replace named import 1`] = `
"import {MuiIcons} from \\"@sui/deps-material\\";

const a = <MuiIcons.DeleteForever attr=\\"1\\">2</MuiIcons.DeleteForever>;
const b = <MuiIcons.DeleteForever attr=\\"1\\"/>;
const c = <DrawerVisibleButton
            edge=\\"start\\"
            disableRotate={isMobile}
            icon={isMobile && <MuiIcons.DeleteForever/>}
          />;
const d = MuiIcons.DeleteForever as any;"
`;
